app/Platform/Services/UserAgentService.php

```diff
@@ -163,4 +164,101 @@ private function trimOperatingSystem(string $os): string
     {
         return trim(strtok($os, ';'));
     }
+
+    public static function versionCompare(string $versionOne, string $versionTwo): int
+    {
+        $versionOneParts = UserAgentService::splitVersion($versionOne);
+        $versionOnePartCount = count($versionOneParts);
+        $versionTwoParts = UserAgentService::splitVersion($versionTwo);
+        $versionTwoPartCount = count($versionTwoParts);
+
+        $index = 0;
+        while (true) {
+            if ($index === $versionOnePartCount) {
+                return ($index === $versionTwoPartCount) ? 0 : -1;
+            }
+
+            if ($index === $versionTwoPartCount) {
+                return 1;
+            }
+
+            $versionOnePart = $versionOneParts[$index];
+            $versionTwoPart = $versionTwoParts[$index];
+            $diff = strcmp($versionOnePart, $versionTwoPart);
+            if ($diff !== 0) {
+                $versionOneNumeric = ctype_digit($versionOnePart);
+                $versionTwoNumeric = ctype_digit($versionTwoPart);
+                if ($versionOneNumeric) {
+                    if ($versionTwoNumeric) {
+                        // both parts are fully numeric. compare them numerically
+                        $versionOneNumber = (int) $versionOnePart;
+                        $versionTwoNumber = (int) $versionTwoPart;
+                        if ($versionOneNumber < $versionTwoNumber) {
+                            return -1;
+                        } elseif ($versionOneNumber > $versionTwoNumber) {
+                            return 1;
+                        } else {
+                            return 0;
+                        }
+                    }
+
+                    // 1.0.1 > 1.0-dirty
+                    return 1;
+                } elseif ($versionTwoNumeric) {
+                    // 1.0-dirty < 1.0.1
+                    return -1;
+                }
+
+                // neither part is fully numeric. compare them as strings
+                return ($diff < 0) ? -1 : 1;
+            }
+
+            $index++;
+        }
+    }
+
+    private static function splitVersion(string $version): array
```

Comment:
What about a case like "1.2.3-beta.1" and "1.2.3-beta.2"? Not sure how this is handled currently - this may already be covered.

Reply:
This is not handled. Both versions will be normalized to "1.2.3.[extra]" and sort after "1.2.3", but be equivalent to each other. It's a similar problem to "1.2.3-branch1" and "1.2.3-branch2", where the order can't really be determined, we just know they're newer than "1.2.3".

Reply:
Alpha/beta versions may generically problematic in any scenario, as "1.2-beta" should presumably be _before_ "1.2".